### PR TITLE
Prevent callFrame and value stack from allocated on gorutine stack

### PR DESCRIPTION
**Pull request by mathetake**
Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

**Comment by mathetake** [comment]
I hope the comments on the change explain well!

File: internal/wasm/jit/engine.go [hunk 1]
```diff
@@ -468,8 +468,30 @@ func newEngine() *engine {
...
 }
 
-// TODO: better make them configurable?
-const (
+// Do not make these variables as constants, otherwise there would be
+// dangerous memory accees from native code.
+//
+// Background: Go has a mechanism called "goroutine stack-shrink" where Go
+// runtime shrinks Gorotuine's stack when it is GCing. Shrinking means that
+// all the contents on the goroutine stack will be relocated by runtime,/
```
Comment (codefromthecrypt) [inline internal/wasm/jit/engine.go]:
```suggestion
// all the contents on the goroutine stack will be relocated by runtime.
```

File: internal/wasm/jit/engine.go [hunk 2]
```diff
@@ -468,8 +468,30 @@ func newEngine() *engine {
...
+// therefore the memory address of these contents change undeterministically.
```
Comment (codefromthecrypt) [inline internal/wasm/jit/engine.go:477]:
```suggestion
// Therefore, the memory address of these contents change undeterministically.
```

File: internal/wasm/jit/engine.go [hunk 3]
```diff
@@ -468,8 +468,30 @@ func newEngine() *engine {
...
+// On the other hand, we hold pointers to the data region of value stack and
+// callframe stack slices and use these raw pointers from  native code.
```
Comment (codefromthecrypt) [inline internal/wasm/jit/engine.go:482]:
```suggestion
// callframe stack slices and use these raw pointers from native code.
```

File: internal/wasm/jit/engine.go [hunk 4]
```diff
@@ -468,8 +468,30 @@ func newEngine() *engine {
...
+// Therefore, it is dangerous if these two stacks are allocated on stack
+// as these stack's address might be changed by Goruntime which we cannot
+// detect.
+//
+// By declaring thses values as values, slices created via make([]..., var) will
```
Comment (codefromthecrypt) [inline internal/wasm/jit/engine.go:487]:
```suggestion
// By declaring theses values as `var`, slices created via `make([]..., var)`will
```

File: internal/wasm/jit/engine.go [hunk 5]
```diff
@@ -468,8 +468,30 @@ func newEngine() *engine {
...
+// never be allocated on stack [1], therefore we can ensure that accessing these
```
Comment (codefromthecrypt) [inline internal/wasm/jit/engine.go:488]:
```suggestion
// never be allocated on stack [1]. This means accessing these
```

File: internal/wasm/jit/engine.go [hunk 6]
```diff
@@ -468,8 +468,30 @@ func newEngine() *engine {
...
+// slices via raw pointers is safe because Go's GC never relocate heap-allocated
```
Comment (codefromthecrypt) [inline internal/wasm/jit/engine.go:489]:
```suggestion
// slices via raw pointers is safe: As of version 1.18, Go's garbage collectornever relocates heap-allocated
```

File: internal/wasm/jit/engine.go [hunk 7]
```diff
@@ -468,8 +468,30 @@ func newEngine() *engine {
...
+// objects (aka no compation of memory [2]). We can actually check how go compiler
```
Comment (codefromthecrypt) [inline internal/wasm/jit/engine.go:490]:
```suggestion
// objects (aka no compilation of memory [2]). On Go upgrades, re-validate via
// `go build -gcflags='-m' ./internal/wasm/jit/...`.
```

File: internal/wasm/jit/engine.go [hunk 8]
```diff
@@ -468,8 +468,30 @@ func newEngine() *engine {
...
+// does escape analysis on these variables lile 'go build -gcflags='-m' ./internal/wasm/jit/...'.
```
Comment (codefromthecrypt) [inline internal/wasm/jit/engine.go:491]:
```suggestion
```

**Review by codefromthecrypt — APPROVED** [review]
clever detective work and thanks for adding the test!

**Pull request merged**